README
Conic Planet 1.2
July 17, 2000

Conic Planet is based on and similar to XPlanet, where an image of
the earth is rendered.  Azimuthal, Mercator, Mollweide, orthographic, or
rectangular projections can be displayed as well as a window with a
globe the user can rotate interactively.  The other terrestrial
planets may also be displayed.  The latest version can be found
by visiting the Conic Planet project page at:
http://sourceforge.net/projects/conic/

A lot of code of Conic Planet has been written by Hari Nair
<hari@alumni.caltech.edu>. So a lot of credit for this work
goes to him.

Most of the algorithms are taken from "Astronomical Formulae for
Calculators" by Jean Meeus, published by Willman-Bell.  The rotational
parameters for the other terrestrial planets are taken from "Practical
Ephemeris Calculations" by Oliver Montenbruck, published by
Springer-Verlag.  Rotational parameters for the Moon are taken from
Davies et al. (1996), Celestial Mechanics 63, 127--148.  Most of the
map projection algorithms are from "Map Projections - A Working
Manual" by John P. Snyder, published by the U.S. Department of the
Interior.

Conic Planet is free software, distributed according to the terms of the
GNU General Public License.  See the COPYING file for details.

See the INSTALL file for installation instructions.


--------------------------------------------------------------


Valid CGI key/value pair options to Conic Planet are:

body=mercury/venus/earth/moon/mars/jupiter/io/europa/ganymede/callisto/
     saturn/titan/uranus/neptune/pluto
Render an image of the specified planet/moon.

centerx=xvalue
centery=yvalue
Place the center of the globe at (xvalue,yvalue). You can use this with
the "radius" option to put a small image anywhere on the screen.  If no
projection is explicitly specified, orthographic is assumed, but this
option may also be used with the azimuthal and mollweide projections.

clouds=0/1 [0 = no, 1 = yes] - currently unimplemented
Overlay clouds onto image. Overlaying clouds slows Conic Planet
down considerably (but it looks really nice).

position=dayside/earthside/moonside/nightside

	dayside:
Render the image as seen from directly above the subsolar point.  If no
projection is explicitly specified, orthographic is assumed, but this
option may also be used with any projection except rectangular. 

	earthside:
Render the image as seen from the earth. This option only works with
other planets, specified with "body".  If no projection is explicitly
specified, orthographic is assumed, but this option may also be used
with any projection except rectangular.

	moonside:
Render the image as seen from the moon.  If no projection is
explicitly specified, orthographic is assumed, but this option may
also be used with any projection except rectangular.

	nightside                   
Render the image as seen from directly above the anti-subsolar point.
If no projection is explicitly specified, orthographic is assumed, but
this option may also be used with any projection except rectangular.

filter=0/1 [0 = no, 1 = yes]
Use bilinear interpolation to compute the color of each pixel instead
of nearest-neighbor interpolation.  It slows down the computation, but
it can look a lot better, particularly if you're using a low
resolution map.  If no projection is explicitly specified,
orthographic is assumed, but this option may also be used with any
projection except rectangular.

fuzz=fuzzfactor
Let the day and night hemispheres blend into one another for pixels
within fuzz degrees of the terminator.  The default value is 6.

width=widthvalue
height=heightvalue
Specify the width and height of the output bitmap.  

grid=0/1  [0 = no, 1 = yes]
Draw a longitude/latitude grid.  The spacing of major grid lines and
dots between major grid lines can be controlled with the "grid1" and
"grid2" options (see below).

grid1=grid1
Specify the spacing of grid lines. Grid lines are drawn with a
90/grid1 degree spacing. The default value for grid1 is 6,
corresponding to 15 degrees between major grid lines.  This option
implies the "grid" option.

grid2=grid2
Specify the spacing of dots along grid lines.  Grid dots are drawn
with a 90/(grid1 x grid2) degree spacing.  The default value for grid2
is 15; combined with the default grid1 value of 6, this corresponds to
placing grid dots on a one degree spacing.  This option implies the
"grid" option.

latitude=latitudevalue
Render the globe as seen from above the specified latitude (in
degrees).  The default value is 0. If no projection is
explicitly specified, orthographic is assumed, but this option may
also be used with any projection except rectangular.

longitude=longitudevalue
Place the observer above the specified longitude (in degrees).
Longitude is positive going east, negative going west (for earth and
moon), so for example Los Angeles is at -118 or 242.  The default
value is 0.

night=0/1 [0 = no, 1 = yes]
Use a night map image.  If this option is not
specified, the night map will be a copy of the day map,
modified as described with the "shade" parameter.

projection=azimuthal/mercator/mollweide/orthographic/rectangular
The projection type may be one of azimuthal, mercator, mollweide,
orthographic, or rectangular.

radius=radiusvalue
Render the globe with a radius of radiusvalue percent of the screen height.
The default value is 50% of the screen height.  If no projection is
explicitly specified, orthographic is assumed, but this option may
also be used with the azimuthal and mollweide projections.  If used
with the mollweide projection, the radius value is the value of the
semimajor (horizontal) axis as a percent of the screen width.

range=rangevalue
Render the globe as seen from a distance of rangevalue from the planet's
center, in units of the planetary radius.  The default value is 1000.
Note that if you use very close ranges the field of view of the screen
can be a lot greater than 180 degrees!  If you want an "up close" image
use the "radius" option.  This option implies projection=orthographic. 

rotate=angle 
Rotate the globe by angle degrees counterclockwise so that north isn't
at the top.  The default value is 0.  My friends in the Southern
Hemisphere can use rotate=180 to make the earth look like it should!
If no projection is explicitly specified, orthographic is assumed, but
this option may also be used with any projection except rectangular.
For non-orthographic projections, the globe is rotated and then
projected, if that helps you visualize what to expect.

shade=shadevalue
If night=1 is not specified, set the brightness of the night map
to shade percent of the day map.  If shade is 100, the day and night
maps will be identical.  The default value is 30.

smoothen=factor [1 = off, 2 = double image size, 3 = triple image size, ...]
This option multiplies the image resolution by a factor, and then
uses bilinear interpolation to resize the image to the actual size.
It slows down the computation, but it can look a lot better,
particularly if you're generating low res pictures.

starfreq=frequency
Fraction of background pixels that will be colored white.  The default
value is 0.001. If no projection is explicitly specified, orthographic
is assumed, but this option may also be used with the azimuthal and
Mollweide projections.

terminator=morning/evening
Place the observer above the specified terminator. For
non-rectangular projections, the image will be rotated so that
the terminator is approximately vertical.  This can be combined
with the "rotate" option to orient the terminator any way you want.


Conic Planet looks in the directory specified in auxfiles.h
for the file to use as the day map.  This file should be named
body.jpg, where body can be any of the various body names such
as earth.jpg, mars.jpg, etc. The night-map is called body-night.jpg,
such as earth-night.jpg. The cloud-map takes the form body-cloud.png,
such as earth-cloud.png (Note that this is in PNG format).

Mukund Iyer
muks@crosswinds.net



ChangeLog

ChangeLog for Conic Planet:

17 July 2000

	Version 1.2
	Added code for enhancment of contrast of the SSEC cloud image
	and smoothening of seam at 180 deg. longitude line in the
	SSEC cloud image. Added brightness and gamma-correction
	controls. Added support for greyscale (R=G=B) output.

5 July 2000

	Version 1.1.1f
	Bug fix for a large error in scaling code.

5 July 2000

	Version 1.1.1
	Cloud handling code modified to include a threshold
	level for what clouds are shown. The option "smoothen"
	was added for generating image at a higher resolution and
	using a bilinear filter to resize it, so that aliasing artifacts
	are smoothened out. PNG support was added for output content-type.
	The output content-type is now set using the "contenttype" option.
	Some other minor tweaks.

27 June 2000

	Version 1.1.0
	Support for more planets, moons n' stuff. Another option
	"terminator" added. Clouds support added using the
	cloud information available at the Space Science and
	Engineering Center (SSEC) at the University of Wisconsin.
	Many thanks again to Hari Nair (I swear I tried it myself,
	but that border killed my mollweide->rectangular
	transformation). Added aclocal.m4 for detecting libjpeg.
	A PHP form (test.php) is included to view the program's various
	features. This can be used with a PHP-enabled webserver.

14 June 2000

	Version 1.0.0
	Initial public code-release


Mukund Iyer
muks@crosswinds.net



planet.cc
/******************************************************************************
    Conic Planet - CGI program to render an image of a planet
    Copyright (C) 2000 Mukund Iyer <muks@crosswinds.net>
    
    Most portions of this code were developed by and are
    Copyright (C) 2000 Hari Nair <hari@alumni.caltech.edu>
    for XPlanet 0.72

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
******************************************************************************/

/*
  planet.cc - parameters for orbital and rotational motion.  Orbital
  parameters are from Chapter 23 of Astronomical Formulae for
  Calculators by Meeus, rotational parameters for other planets and
  satellites are from Davies et al. (1996), Celestial Mechanics 63,
  127--148.  

  Formulae for Pluto are taken from Paul Schlyter's page at 
  http://www.m2c3.com/alpocs/tdl1999/schlyter110199/ppcomp.html
*/

#include <iostream>
#include <math.h>
#include <stdlib.h>

#include "options.h"
#include "planet.h"
#include "util.h"

int flipped = 1;  
/*
flipped = 1 if planet's longitude increases to the east (e.g. earth),
         -1 if planet's longitude increases to the west (e.g. Mars),
*/

char *body[] = {"mercury", "venus", "earth", "moon", "mars",
                          "jupiter", "io", "europa", "ganymede", "callisto",
                          "saturn", "titan", "uranus", "neptune",
                          "pluto"};
                                                                              

planet::planet (double T)
{
  double T2000 = T - 1;                      // Centuries from 2000.0
  double d = T2000 * 36525;                  // Days from 2000.0

  int i;
      
  switch (opts.body)
    {
    case MERCURY:
      mlong = (poly (1.78179078E2, 1.4947407078E5, 3.011E-4, 0, T)
	       * deg_to_rad);
      axis = 0.3870986;
      ecc = poly (2.0561421E-1, 2.046E-5, -3E-8, 0, T);
      incl = (poly (7.002881, 1.8608E-3, -1.83E-5, 0, T)
	      * deg_to_rad);
      perih = (poly (2.8753753E1, 3.702806E-1, 1.208E-4, 0, T)
	       * deg_to_rad);
      ascnode = (poly (4.7145944E1, 1.1852083, 1.739E-4, 0, T)
		 * deg_to_rad);
      flipped = -1; 

      alpha0 = 281.01 - 0.033 * T2000;
      delta0 = 61.45 - 0.005 * T2000;

      null_meridian = 329.68;
      wdot = 6.1385025;

      break;
    case VENUS:
      mlong = (poly (3.42767053E2, 5.851921191E4, 3.097E-4, 0, T)
	       * deg_to_rad);
      axis = 0.7233316;
      ecc = poly (6.82069E-3, -4.774E-5, 9.1E-8, 0, T);
      incl = (poly (3.393631, 1.0058E-3, -1.0E-6, 0, T)
	      * deg_to_rad);
      perih = (poly (5.4384186E1, 5.081861E-1, -1.3864E-3, 0, T)
	       * deg_to_rad);
      ascnode = (poly (7.5779647E1, 8.9985E-1, 4.1E-4, 0, T)
		 * deg_to_rad);
      flipped = 1; 

      alpha0 = 272.76;
      delta0 = 67.16;

      null_meridian = 160.20;
      wdot = -1.4813688;
      
      break;
    case MOON:
      double E[14];
      E[ 1] = 125.045 -  0.0529921 * d;
      E[ 2] = 250.089 -  0.1059842 * d;
      E[ 3] = 260.008 + 13.0120009 * d;
      E[ 4] = 176.625 + 13.3407154 * d;
      E[ 5] = 357.529 +  0.9856003 * d;
      E[ 6] = 311.589 + 26.4057084 * d;
      E[ 7] = 134.963 + 13.0649930 * d;
      E[ 8] = 276.617 +  0.3287146 * d;
      E[ 9] =  34.226 +  1.7484877 * d;
      E[10] =  15.134 -  0.1589763 * d;
      E[11] = 119.743 +  0.0036096 * d;
      E[12] = 239.961 +  0.1643573 * d;
      E[13] =  25.053 + 12.9590088 * d;
      for (i = 1; i < 14; i++) E[i] *= deg_to_rad;

      alpha0 = 269.9949 + (0.0031 * T2000         - 3.8787 * sin (E[ 1])
			   - 0.1204 * sin (E[ 2]) + 0.0700 * sin (E[ 3])
			   - 0.0172 * sin (E[ 4]) + 0.0072 * sin (E[ 6])
			   - 0.0052 * sin (E[10]) + 0.0043 * sin (E[13]));
      delta0 =  66.5392 + (0.0130 * T2000         + 1.5419 * cos (E[ 1])
			   + 0.0239 * cos (E[ 2]) - 0.0278 * cos (E[ 3])
			   + 0.0068 * cos (E[ 4]) - 0.0029 * cos (E[ 6])
			   + 0.0009 * cos (E[ 7]) + 0.0008 * cos (E[10])
			   - 0.0009 * cos (E[13]));
      null_meridian = 38.3213 + (d * (13.17635815 - 1.4E-12 * d) 
				 + 3.5610 * sin (E[ 1]) + 0.1208 * sin (E[ 2])
				 - 0.0642 * sin (E[ 3]) + 0.0158 * sin (E[ 4])
				 + 0.0252 * sin (E[ 5]) - 0.0066 * sin (E[ 6])
				 - 0.0047 * sin (E[ 7]) - 0.0046 * sin (E[ 8])
				 + 0.0028 * sin (E[ 9]) + 0.0052 * sin (E[10])
				 + 0.0040 * sin (E[11]) + 0.0019 * sin (E[12])
				 - 0.0044 * sin (E[13]));
      wdot = 0;
      flipped = 1;

      break;
    case MARS:
      mlong = (poly (2.93737334E2, 1.914169551E4, 3.107E-4, 0, T) 
	       * deg_to_rad);
      axis = 1.5236883;
      ecc = poly (9.33129E-2, 9.2064E-5, -7.7E-8, 0, T);
      incl = (poly (1.850333, -6.75E-4, 1.26E-5, 0, T) * deg_to_rad);
      perih = (poly (2.85431761E2, 1.0697667, 1.313E-4, 4.41E-6, T) 
	       * deg_to_rad);
      ascnode = (poly (4.8786442E1, 7.709917E-1, -1.4E-6, -5.33E-6, T) 
		 * deg_to_rad);
      flipped = -1;
      
      alpha0 = 317.681 - 0.108 * T2000;
      delta0 = 52.886 - 0.061 * T2000;
      null_meridian = 176.901;
      wdot = 350.891983;
      
      break;
    case JUPITER:
    case IO:
    case EUROPA:
    case GANYMEDE:
    case CALLISTO:
      mlong = (poly (2.38049257E2, 3.036301986E3, 3.347E-4, -1.65E-6, 
		     T) * deg_to_rad);
      axis = 5.202561;
      ecc = poly (4.833475E-2, 1.6418E-4, -4.676E-7, -1.7E-9, T);
      incl = poly (1.308736, -5.6961E-3, 3.9E-6, 0, T) * deg_to_rad;
      perih = (poly (2.73277558E2, 5.994317E-1, 7.0405E-4, -5.08E-6, T) 
	       * deg_to_rad);
      ascnode = (poly (9.9443414E1, 1.01053, 3.5222E-4, -8.51E-6, T) 
		 * deg_to_rad);

      flipped = -1;

      double J[9];
      J[1] =  73.32 + 91472.9 * T2000;
      J[2] =  24.62 + 45137.2 * T2000;
      J[3] = 283.90 +  4850.7 * T2000;
      J[4] = 355.80 +  1191.3 * T2000;
      J[5] = 119.90 +   262.1 * T2000;
      J[6] = 229.80 +    64.3 * T2000;
      J[7] = 352.25 +  2382.6 * T2000;
      J[8] = 113.35 +  6070.0 * T2000;
      for (i = 1; i < 9; i++) J[i] *= deg_to_rad;

      switch (opts.body)
	{
	case JUPITER:
	  alpha0 = 268.05 - 0.009 * T2000;
	  delta0 = 64.49 + 0.003 * T2000;
	  
	  // System III (magnetic field rotation)
	  null_meridian = 284.95;
	  wdot = 870.536;
	  break;
	case IO:
	  alpha0 = (268.05 - 0.009 * T2000 + 0.094 * sin (J[3]) 
		    + 0.024 * sin (J[4]));
	  delta0 = (64.50 + 0.003 * T2000 + 0.040 * cos (J[3])
		    + 0.011 * cos (J[4]));
	  null_meridian = 200.39 - 0.085 * sin (J[3]) - 0.022 * sin (J[4]);
	  wdot = 203.4889538;
	  break;
	case EUROPA:
	  alpha0 = (268.08 - 0.009 * T2000 + 1.086 * sin (J[4])
		    + 0.060 * sin (J[5]) + 0.015 * sin (J[6])
		    + 0.009 * sin (J[7]));
	  delta0 = (64.51 + 0.003 * T2000 + 0.468 * cos (J[4])
		    + 0.026 * cos (J[5]) + 0.007 * cos (J[6])
		    + 0.002 * cos (J[7]));
	  null_meridian = (35.67 - 0.980 * sin (J[4]) - 0.054 * sin (J[5])
			   - 0.014 * sin (J[6]) - 0.008 * sin (J[7]));
	  wdot = 101.3747235;
	  break;
	case GANYMEDE:
	  alpha0 = (268.20 - 0.009 * T2000 - 0.037 * sin (J[4]) 
		    + 0.431 * sin (J[5]) + 0.091 * sin (J[6]));
	  delta0 = (64.57 + 0.003 * T2000 - 0.016 * cos (J[4])
		    + 0.186 * cos (J[5]) + 0.039 * cos (J[6]));
	  null_meridian = (44.04 + 0.033 * sin (J[4]) - 0.389 * sin (J[5])
			   - 0.082 * sin (J[6]));
	  wdot = 50.3176081;
	  break;
	case CALLISTO:
	  alpha0 = (268.72 - 0.009 * T2000 - 0.068 * sin (J[5])
		    + 0.590 * sin (J[6]) + 0.010 * sin (J[8]));
	  delta0 = (64.83 + 0.003 * T2000 - 0.029 * cos (J[5])
		    + 0.254 * cos (J[6]) - 0.004 * cos (J[8]));
	  null_meridian = (259.73 + 0.061 * sin (J[5]) - 0.533 * sin (J[6])
			   - 0.009 * sin (J[8]));
	  wdot = 21.5710715;
	  break;
	default:
	  break;
	}
      break;
    case SATURN:
    case TITAN:
      {
	mlong = (poly (2.66564377E2, 1.223509884E3, 3.245E-4, -5.8E-6, T) 
		 * deg_to_rad);
	axis = 9.554747;
	ecc = poly (5.589232E-2, -3.455E-4, -7.28E-7, 7.4E-10, T);
	incl = (poly (2.492519, -3.9189E-3, -1.549E-5, 4E-8, T) * deg_to_rad);
	perih = (poly (3.383078E2, 1.0852207, 9.7854E-4, 9.92E-6, T) 
		 * deg_to_rad);
	ascnode = (poly (1.12790414E2, 8.731951E-1, -1.5218E-4, -5.31E-6, T)
		   * deg_to_rad);
	flipped = -1;

	double S[8];
	S[7] = 29.80 - 52.1 * T2000;
	S[7] *= deg_to_rad;
	//      for (i = 1; i < 8; i++) S[i] *= deg_to_rad;
	switch (opts.body)
	  {
	  case SATURN:
	    alpha0 = 40.589 - 0.036 * T2000;
	    delta0 = 83.537 - 0.004 * T2000;
	    null_meridian = 38.90;
	    wdot = 810.7939024;
	    break;
	  case TITAN:
	    alpha0 = 36.41 - 0.036 * T2000 + 2.66 * sin (S[7]);
	    delta0 = 83.94 - 0.004 * T2000 - 0.30 * cos (S[7]);
	    null_meridian = 189.64 - 2.64 * sin (S[7]);
	    wdot = 22.576978;
	    break;
	  default:
	    break;
	  }
      }
      break;
    case URANUS:
      mlong = (poly (2.4419747E2, 4.29863546E2, 3.16E-4, -6E-7, T) 
	       * deg_to_rad);
      axis = 19.21814;
      ecc = poly (4.63444E-2, -2.658E-5, 7.7E-8, 0, T);
      incl = (poly (7.72464E-1, 6.253E-4, 3.95E-5, 0, T) * deg_to_rad);
      perih = (poly (9.8071581E1, 9.85765E-1, -1.0745E-3, -6.1E-7, T) 
	       * deg_to_rad);
      ascnode = (poly (7.3477111E1, 4.986678E-1, 1.3117E-3, 0, T) 
		 * deg_to_rad);

      flipped = 1;

      alpha0 = 257.311;
      delta0 = -15.175;
      null_meridian = 203.81;
      wdot = -501.1600928;
      break;
    case NEPTUNE:
      {
	mlong = (poly (8.457994E1, 2.19885914E2, 3.205E-4, 0, T) * deg_to_rad);
	axis = 30.10957;
	ecc = poly (8.99704E-3, 6.33E-7, -2E-9, 0, T);
	incl = (poly (1.779242, -9.5436E-3, -9.1E-6, 0, T) * deg_to_rad);
	perih = (poly (2.76045975E2, 3.256394E-1, 1.4095E-4, 4.113E-6, T) * deg_to_rad);
	ascnode = (poly (1.30681389E2, 1.098935, 2.4987E-4, -4.718E-6, T) * deg_to_rad);
	
	flipped = -1;
	double N = (357.85 + 52.316 * T2000) * deg_to_rad;
	alpha0 = 299.36 + 0.70 * sin (N);
	delta0 = 43.46 - 0.51 * cos (N);
	null_meridian = 253.18 - 0.48 * sin (N);
	wdot = 536.3128492;
      }
      break;
    case PLUTO:
      {
	double S = ( 50.03 + 0.033459652 * d) * deg_to_rad;
	double P = (238.95 + 0.003968789 * d) * deg_to_rad;
	
	helio.lon = (238.9508 + 0.00400703 * d
		     - 19.799 * sin(P)   + 19.848 * cos(P)
		     +  0.897 * sin(2*P) -  4.956 * cos(2*P)
		     +  0.610 * sin(3*P) +  1.211 * cos(3*P)
		     -  0.341 * sin(4*P) -  0.190 * cos(4*P)
		     +  0.128 * sin(5*P) -  0.034 * cos(5*P)
		     -  0.038 * sin(6*P) +  0.031 * cos(6*P)
		     +  0.020 * sin(S-P) -  0.010 * cos(S-P)) * deg_to_rad;
	
	helio.lat = (-3.9082
		     - 5.453 * sin(P)   - 14.975 * cos(P)
		     + 3.527 * sin(2*P) + 1.673 * cos(2*P)
		     - 1.051 * sin(3*P) + 0.328 * cos(3*P)
		     + 0.179 * sin(4*P) - 0.292 * cos(4*P)
		     + 0.019 * sin(5*P) + 0.100 * cos(5*P)
		     - 0.031 * sin(6*P) - 0.026 * cos(6*P)
		     + 0.011 * cos(S-P)) * deg_to_rad;
	
	helio.dist = (40.72
		      + 6.68 * sin(P)   + 6.90 * cos(P)
		      - 1.18 * sin(2*P) - 0.03 * cos(2*P)
		      + 0.15 * sin(3*P) - 0.14 * cos(3*P));
	
	alpha0 = 313.02;
	delta0 = 9.09;
	null_meridian = 236.77;
	wdot = -56.3623195;
	
	flipped = 1;
      }
      break;
    default:
      break;
    }
  
  alpha0 *= deg_to_rad;
  delta0 *= deg_to_rad;
  null_meridian += wdot * d;
  null_meridian = fmod (null_meridian, 360.0);
  null_meridian *= deg_to_rad;  

  name = body[opts.body];
}

void 
planet::calc_helio ()
{
  if (opts.body == PLUTO) return;
  manom = mlong - perih - ascnode;
  double eccanom = kepler (ecc, manom);
  double nu = 2 * atan (sqrt ((1 + ecc) / (1 - ecc)) 
			* tan (eccanom/2));
  double lat = perih + nu;
  helio.lon = atan2 (cos (incl) * sin (lat), cos (lat)) + ascnode;
  helio.lat = asin (sin (lat) * sin (incl));
  helio.dist = axis * (1 - ecc * cos (eccanom));
}

void 
planet::calc_geo (const double sunlon, const double sundist)
{
  double N = helio.dist * cos (helio.lat) * sin (helio.lon - sunlon);
  double D = (helio.dist * cos (helio.lat) * cos (helio.lon - sunlon)
	      + sundist);
  geo.dist = N*N + D*D + 
    (helio.dist * sin (helio.lat)) * (helio.dist * sin (helio.lat));
  geo.dist = sqrt (geo.dist);
  geo.lon = atan2 (N, D) + sunlon;
  geo.lat = asin (helio.dist * sin (helio.lat) / geo.dist);
}

coordinates
planet::calc_coordinates (coordinates coords, double eps)
{
  compute_ra_dec (coords.lon, coords.lat, coords.alpha, coords.delta, eps);
  /*
  cout << "inclination = " << incl / deg_to_rad << '\n' 
       << "longitude of ascending node = " << ascnode / deg_to_rad << '\n' 
       << "longitude of perihelion = " << perih / deg_to_rad << '\n' 
       << "mean distance = " << axis << '\n' 
       << "eccentricity = " << ecc << '\n' 
       << "mean longitude = " << mlong / deg_to_rad << endl;
  
  cout << "heliocentric longitude = " << helio.lon / deg_to_rad << ", "
       << "heliocentric latitude = " << helio.lat / deg_to_rad << '\n'
       << helio.dist << endl;
  
  cout << "RA = " << coords.alpha << ", " 
       << "dec = " << coords.delta / deg_to_rad << endl;
  cout << "North pole alpha = " << alpha0/deg_to_rad << ", delta = " 
       << delta0/deg_to_rad << endl;
  */
  coords.alpha = TWO_PI * coords.alpha / 24.;

  axis_position = atan2 (cos (delta0) * sin (alpha0 - coords.alpha),
			 sin (delta0) * cos (coords.delta) 
			 - (cos (delta0) * sin (coords.delta) 
			    * cos (alpha0 - coords.alpha)));
  double lat = -asin (sin (delta0) * sin (coords.delta) 
		      + (cos (delta0) * cos (coords.delta) 
			 * cos (alpha0 - coords.alpha)));
  double lon = atan2 (sin (delta0) * cos (coords.delta) 
		      * cos (alpha0 - coords.alpha)
		      - cos (delta0) * sin (coords.delta),
		      cos (coords.delta) * sin (alpha0 - coords.alpha));

  double light_time = 5.7756E-3 * geo.dist * wdot * deg_to_rad;
  lon -= null_meridian;
  lon = fmod (lon, TWO_PI);
  if (lon < 0) lon += TWO_PI;
  coordinates returnval = {lon, lat, axis_position, light_time, 0};
  return (returnval);	
}
